[Merged by Bors] - feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

strihanje01 opened the pull request:
add theorem about the structure of a doubling smaller than the golden ratio


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [x] depends on: #abc [optional extra text]

-->
- [x] depends on: #28653

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): add weak non-commutive Kneser's theorem

add theorem about the structure of a doubling smaller than the golden ratio

strihanje01 commented:
@YaelDillies here's the PR for the Kneser's theorem

YaelDillies reviewed (commented):
Thank you very much for the PR! Here is a first round of very generic comments.

Would you be interested in tackling [the "generalisation" to doubling `2 - ε`](https://terrytao.wordpress.com/2011/03/12/hamidounes-freiman-kneser-theorem-for-nonabelian-groups/#more-4698) at some later point?

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

resolve the first round of comments

strihanje01 commented:
-awaiting-author

strihanje01 commented:
> Thank you very much for the PR! Here is a first round of very generic comments.
> 
> Would you be interested in tackling [the "generalisation" to doubling `2 - ε`](https://terrytao.wordpress.com/2011/03/12/hamidounes-freiman-kneser-theorem-for-nonabelian-groups/#more-4698) at some later point?

Yes, I would like to do the case 2-ε.

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+open goldenRatio in
+/-- If `A` has doubling strictly less than `φ`, then `A * A⁻¹` is covered by
+at most constant number of cosets of a finite subgroup of `G`. -/
+theorem doubling_lt_golden_ratio {K : ℝ} (hK1 : 1 < K) (hK2 : K < φ)
```
```suggestion
theorem doubling_lt_golden_ratio {K : ℝ} (hK₁ : 1 < K) (hKφ : K < φ)
```
is slightly better, maybe

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+at most constant number of cosets of a finite subgroup of `G`. -/
+theorem doubling_lt_golden_ratio {K : ℝ} (hK1 : 1 < K) (hK2 : K < φ)
+    (hA1 : #(A⁻¹ * A) ≤ K * #A) (hA2 : #(A * A⁻¹) ≤ K * #A) :
+    ∃ (H : Subgroup G) (_ : Fintype H) (Z : Finset G) (_: #Z ≤ K * (2-K) / (1-K*(K-1))),
```
```suggestion
    ∃ (H : Subgroup G) (_ : Fintype H) (Z : Finset G) (_: #Z ≤ K * (2 - K) / (1-K*(K-1))),
```
and everywhere else

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    ∃ (H : Subgroup G) (_ : Fintype H) (Z : Finset G) (_: #Z ≤ K * (2-K) / (1-K*(K-1))),
+    A * A⁻¹ = (H : Set G) * Z := by
+  classical
+
+  have K_pos : 0 < K := by positivity
```
It's a nice pattern to prove all those positivity statements at once
```suggestion
  have hK₀ : 0 < K := by positivity
  have hKφ' : 0 < φ - K := by linarith
  have hKψ' : 0 < K - ψ := by linarith [goldConj_neg]
  have hK₂' : 0 < 2 - K := by linarith [gold_lt_two]
  have const_pos : 0 < K * (2 - K) / ((φ - K) * (K - ψ)) := by positivity
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+          ← sub_add, gold_mul_goldConj, sub_neg_eq_add]
+    ring_nf
+  rw [calc1]
+
+  have const_pos : K * (2-K) / ((φ - K) * (K - ψ)) > 0 := by
+    apply div_pos
+    · apply mul_pos
+      · linarith
+      · linarith [hK2, gold_lt_two]
+    · apply mul_pos
+      · linarith [hK2]
+      · linarith [hK1, goldConj_neg]
+
```
```suggestion

```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+      · linarith [hK1, goldConj_neg]
+
+  obtain rfl | emptyA := A.eq_empty_or_nonempty
+  · refine ⟨⊥, inferInstance, ∅, ?_, ?_⟩
+    · simp [card_empty, CharP.cast_eq_zero]
+      linarith only [const_pos]
+    · simp
+
```
```suggestion
  · exact ⟨⊥, inferInstance, ∅, by simp [const_pos.le]⟩
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    · simp
+
+  let S := A * A⁻¹
+  have id_in_S : 1 ∈ S := by
+    obtain ⟨a, ha⟩ := emptyA
+    rw [← mul_inv_cancel a]
+    exact mul_mem_mul ha (inv_mem_inv ha)
+
```
```suggestion
  have id_in_S : 1 ∈ S := by obtain ⟨a, ha⟩ := emptyA; simpa using mul_mem_mul ha (inv_mem_inv ha)
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    exact mul_mem_mul ha (inv_mem_inv ha)
+
+  let H := stabilizer G S
+
```
```suggestion

```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    simpa [H, ← stabilizer_coe_finset] using stabilizer_finite (by simpa [S]) S.finite_toSet
+  cases nonempty_fintype H
+
+  use H
+  use inferInstance
+
+  let preZ := {cH ∈ orbit Gᵐᵒᵖ (H : Set G) | Set.Nonempty ((cH : Set G) ∩ (S: Set G))}
```
```suggestion
  let preZ := {cH ∈ orbit Gᵐᵒᵖ (H : Set G) | (cH ∩ S : Set G).Nonempty}
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    exact Finite.Set.finite_of_finite_image preZ inj_f_preZ
+
+  let preZ' := fin_preZ.toFinset
+  have elts_preZ'_nonempty {cH : Set G} : (cH ∈ preZ') → Set.Nonempty (cH ∩ S) := by
+    intro hcH
```
```suggestion
  have elts_preZ'_nonempty (cH : Set G) (hcH : cH ∈ preZ') : (cH ∩ S).Nonempty := by
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    intro hcH
+    rw [Set.Finite.mem_toFinset, Set.mem_setOf] at hcH
+    exact hcH.2
+
+  let chooseZ : {cH : Set G | cH ∈ preZ'} → G := fun cH =>
+    Classical.choose (elts_preZ'_nonempty cH.2)
+
+  let Z := Finset.image chooseZ preZ'.attach
+
```
The `choose` tactic is very useful for these sorts of things
```suggestion
  choose! chooseZ chooseZ_spec using elts_preZ'_nonempty
  let Z := Finset.image chooseZ preZ'
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+  let Z := Finset.image chooseZ preZ'.attach
+
+  have Z_subset_S : Z ⊆ S := by
+    intro z hz
+    unfold Z at hz
+    rw [mem_image] at hz
+    obtain ⟨c, hc1, hc2⟩ := hz
+    unfold chooseZ at hc2
+    have hc : (c : Set G) ∈ preZ' := by exact coe_mem c
+    rw [Set.Finite.mem_toFinset fin_preZ] at hc
+    apply And.right at hc
+    apply Classical.choose_spec at hc
+    apply Set.mem_of_mem_inter_right at hc
+    simp_all only [setOf_mem, coe_sort_coe, mem_attach, mem_coe]
```
```suggestion
    simpa [Z, Finset.image_subset_iff] using fun x hx ↦ (chooseZ_spec _ hx).2
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    · calc
+        (H: Set G) * Z  ⊆ H * S   := by gcongr
+                      _ ⊆ S       := by exact Eq.subset HS_eq_S
+
```
Might as well do all the data-providing in one go:
```suggestion

  refine ⟨H, inferInstance, Z, ?_, mod_cast S_eq_HZ⟩

```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+      obtain ⟨y, hy, z, hz, hyz⟩ := hxz
+      rw [← mem_coe, S_eq_HZ, ← hyz]
+      exact Set.mul_mem_mul (Set.mem_toFinset.mp hy) (mem_coe.mpr hz)
+
+  refine ⟨Z, ?_, mod_cast S_eq_HZ⟩
```
```suggestion

```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+  refine ⟨Z, ?_, mod_cast S_eq_HZ⟩
+  let r : G → ℕ := fun z => #{xy ∈ A ×ˢ A | xy.1 * (xy.2 : G)⁻¹ = z}
+
+  have many_big_z : #{z ∈ S | r z > (K - 1) * #A} ≥ (φ-K)*(K-ψ)/(2-K) * #A := by
```
```suggestion
  have many_big_z : #{z ∈ S | r z > (K - 1) * #A} ≥ (φ-K)*(K-ψ)/(2 - K) * #A := by
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+  have many_big_z : #{z ∈ S | r z > (K - 1) * #A} ≥ (φ-K)*(K-ψ)/(2-K) * #A := by
+    let k := #{z ∈ S | r z > (K - 1) * #A}
+
+    have ineq1 : (#A) * (#A) ≤ ((2 - K) * k + K * (K - 1) * #A) * #A := by
```
```suggestion
    have ineq1 : #A * #A ≤ ((2 - K) * k + K * (K - 1) * #A) * #A := by
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+
+    have ineq1 : (#A) * (#A) ≤ ((2 - K) * k + K * (K - 1) * #A) * #A := by
+      calc
+        (#A : ℝ) * (#A) = #(A ×ˢ A) := by simp only [card_product, Nat.cast_mul]
```
```suggestion
        (#A : ℝ) * #A = #(A ×ˢ A) := by simp only [card_product, Nat.cast_mul]
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+        )
+      case eq2 =>
+        rw [← Nat.cast_add, Nat.cast_inj]
+        exact Eq.symm (sum_filter_add_sum_filter_not S (fun z => r z > (K - 1) * #A) r)
```
```suggestion
        exact Eq.symm (sum_filter_add_sum_filter_not S (fun z => (K - 1) * #A < r z) r)
```
and everywhere else

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+        gcongr with z hz
+        exact not_lt.mp (mem_filter.mp hz).2
+      case eq4 =>
+        apply add_le_add_left
+        rw [card_eq_sum_ones {z ∈ S | ¬↑(r z) > (K - 1) * ↑(#A)}, Nat.cast_sum,
```
```suggestion
        gcongr
        rw [card_eq_sum_ones {z ∈ S | ¬ (K - 1) * #A < r z}, Nat.cast_sum,
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+          rw [inv_inj] at hxy
+          exact Prod.ext h hxy
+      case eq7 =>
+      · rw [add_right_cancel_iff, ← Nat.cast_mul, Nat.cast_inj]
```
```suggestion
        rw [add_right_cancel_iff, ← Nat.cast_mul, Nat.cast_inj]
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+      rw [← calc1]
+      linarith only [ineq2]
+
+    apply le_of_mul_le_mul_of_pos_left ?_ (sub_pos.mpr (lt_trans hK2 gold_lt_two))
```
```suggestion
    apply le_of_mul_le_mul_of_pos_left ?_ hK₂'
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+
+    apply le_of_mul_le_mul_of_pos_left ?_ (sub_pos.mpr (lt_trans hK2 gold_lt_two))
+    have : (2 - K) * ((φ - K) * (K - ψ) / (2 - K) * (#A: ℝ)) = (φ - K) * (K - ψ) * #A := by
+      field_simp [ne_of_gt (sub_pos.mpr (lt_trans hK2 gold_lt_two))]
```
```suggestion
      field_simp [hK₂'.ne']
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    exact ⟨x * y⁻¹ * t, by simp [ht, mul_assoc], ((a * b⁻¹)⁻¹ * t)⁻¹, by simp [ht, mul_assoc]⟩
+
+
+  have big_H : Fintype.card H ≥ (φ-K)*(K-ψ)/(2-K) * #A := by
+    change Fintype.card (H: Set G) ≥ (φ-K)*(K-ψ)/(2-K) * #A
```
```suggestion
  have big_H : Fintype.card H ≥ (φ-K)*(K-ψ)/(2 - K) * #A := by
    change Fintype.card (H: Set G) ≥ (φ-K)*(K-ψ)/(2 - K) * #A
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+      apply Set.mul_mem_mul
+      · simp_all only [Set.mem_toFinset]
+      · exact mem_coe.mpr hxz.2
+    unfold Set.InjOn
+    intro xy hxy wt hwt hxywt
+    rw [mem_coe, mem_product, Set.mem_toFinset, mem_image] at hxy hwt
+    obtain ⟨hx, cx, hcx1, hcx2⟩ := hxy
+    obtain ⟨hw, cw, hcw1, hcw2⟩ := hwt
```
```suggestion

```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+    obtain ⟨hx, cx, hcx1, hcx2⟩ := hxy
+    obtain ⟨hw, cw, hcw1, hcw2⟩ := hwt
+
+    have hcx1 := cx.property
```
```suggestion
    simp only [Set.InjOn, Set.coe_setOf, coe_product, Set.coe_toFinset, coe_image, coe_attach,
      Set.image_univ, Set.mem_prod, SetLike.mem_coe, Set.mem_range, Subtype.exists, and_imp,
      forall_exists_index, Prod.forall, Prod.mk.injEq, Z]
    rintro x y hx hy w t hw ht hxywt
```

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -326,4 +323,454 @@ theorem doubling_lt_three_halves (h : #(A * A) < (3 / 2 : ℚ) * #A) :
+
+  calc
+    #S  ≤ K * #A                                                    := by assumption
+      _ = (φ-K)*(K-ψ)/(2-K) * #A * (K*(2-K)/((φ-K)*(K-ψ)))          := ?_
```
```suggestion
      _ = (φ-K)*(K-ψ)/(2 - K) * #A * (K*(2 - K)/((φ-K)*(K-ψ)))          := ?_
```

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

resolve the second round of comments

YaelDillies reviewed (commented):
Could you please add comments within the proof explaning it? The comments will advantageously replace the line breaks

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

resolve the third round of commets

strihanje01 commented:
-awaiting-author

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

extract independent lemmas from the main proof

YaelDillies commented:
🎉 I think a single PR is fine here

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

extract independent lemmas from the proof of the theorem for doubling strictly less than $\phi$;
add the proof of the theorem for doubling less than $2-\epsilon$

YaelDillies committed:
golf first lemma

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

remove theorem for 2-\epsilon doubling

strihanje01 committed:
Merge branch 'strihanje01_smallDoubling' of github.com:strihanje01/mathlib4 into strihanje01_smallDoubling

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

remove theorem for 2-\epsilon doubling

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

remove the theorem for 2-\epsilon doubling

YaelDillies committed:
big golf

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into strihanje01_smallDoubling

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

add convolution and rewrite rightCosetRepresentingFinset

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

add convolution and rewrite rightCosetRepresentingFinset

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

add convolution and rewrite rightCosetRepresentingFinset

YaelDillies commented on Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:
```diff
@@ -362,7 +362,10 @@ omit [DecidableEq G] in
     x ∈ (H : Set G) <• y ↔ (H : Set G) <• x = H <• y := op_smul_eq_iff_mem (mem_orbit _ _)
 
-def convolution (A B : Finset G) : G → ℕ := fun x => #{ab ∈ A ×ˢ B | ab.1 * ab.2⁻¹ = x}
+/-- Given finite subsets `A` and `B` of a group `G`, convolution of `A` and `B` is a map `G → ℕ`
+that for a given `x ∈ G` returns the number of representations of `x` in the form `x = ab⁻¹`, where
+`a ∈ A`, `b ∈ B`. -/
+private def convolution (A B : Finset G) : G → ℕ := fun x => #{ab ∈ A ×ˢ B | ab.1 * ab.2⁻¹ = x}
```
This should not be private and be put in a new file `Combinatorics.Additive.Convolution` (or something like that)

strihanje01 committed:
feat(Combinatorics/Additive/Convolution): convolution map

add convolution of finite subsets of a group

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

add convolution and rewrite rightCosetRepresentingFinset

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

rewrite rightCosetRepresentingFinset

strihanje01 commented:
-awaiting-author

strihanje01 committed:
Merge remote-tracking branch 'upstream/master' into strihanje01_smallDoubling

strihanje01 committed:
Merge branch 'master' into strihanje01_smallDoubling

strihanje01 committed:
merged code adapted

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

rewrite rightCosetRepresentingFinset using Set.InjOn

strihanje01 committed:
Merge branch 'master' into strihanje01_smallDoubling

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

strihanje01 committed:
Merge branch 'master' into strihanje01_smallDoubling

strihanje01 committed:
feat(Combinatorics/Additive/VerySmallDoubling): weak non-commutative Kneser's theorem

resolve comments

jcommelin reviewed (commented):
Thanks :tada:

bors merge
